File: apps/desktop/src/features/chart/chartOrders.ts
import type {
  ChartOrder,
  ChartOrderDraft,
  ChartOrderKind,
  ChartOrderType,
  OptionContract,
  Position,
} from '@0dtetrader/shared-types';
import { chartOrderCrossed } from '@0dtetrader/shared-types';
import type { ApiClient } from '../../core/api/ApiClient';
import { Store } from '../../core/observable';

interface ChartOrdersState {
  /** Chart symbol currently displayed; lines for other underlyings stay loaded
   *  but are not rendered (the store is the whole account's working set). */
  symbol: string;
  orders: ChartOrder[];
  selectedId: string | null;
  error: string | null;
}

/** Lines that can still fire, and are therefore drawn solid and draggable. */
export function isWorking(order: ChartOrder): boolean {
  return order.status === 'working';
}

/** Lines worth drawing at all — working, plus recently terminal ones so a
 *  failed fire stays on the chart instead of silently vanishing. */
export function isVisible(order: ChartOrder): boolean {
  return order.status === 'working' || order.status === 'triggered' || order.status === 'failed';
}

/**
 * Chart order lines (ChartOrders.swift analog). The server is the source of
 * truth — every mutation is a request, and the response replaces the local row.
 *
 * Triggering is deliberately *not* a local order submission: when the quote
 * stream crosses a level the store asks the server to fire that line, so the
 * client and the server-side watcher take the identical path and one crossing
 * can only ever produce one broker order.
 */
export class ChartOrdersStore extends Store<ChartOrdersState> {
  /** Lines with a trigger request in flight, so a burst of ticks fires once. */
  private readonly firing = new Set<string>();
  /** Bumped per load so a superseded snapshot cannot land after a newer one. */
  private loadSeq = 0;
  /** Ids the socket updated while a load was in flight. The push is strictly
   *  newer than the read that produced the snapshot, so it must win. */
  private readonly pushedDuringLoad = new Set<string>();
  /** Last price seen per underlying, the left end of the crossing segment. */
  private readonly lastPrices = new Map<string, number>();

  constructor(private readonly apiClient: ApiClient) {
    super({ symbol: '', orders: [], selectedId: null, error: null });
  }

  /** Lines drawn on the current chart. */
  get visibleOrders(): ChartOrder[] {
    const { symbol, orders } = this.getState();
    return orders.filter((order) => order.underlying === symbol && isVisible(order));
  }

  byId(id: string): ChartOrder | undefined {
    return this.getState().orders.find((order) => order.id === id);
  }

  setSymbol(symbol: string): void {
    if (symbol === this.getState().symbol) return;
    this.set({ symbol, selectedId: null });
  }

  select(id: string | null): void {
    this.set({ selectedId: id });
  }

  /**
   * Re-reads the account's lines. Called on mount and on every socket
   * reconnect, because pushes that landed while the stream was down are gone
   * for good.
   *
   * Merges rather than replaces. A snapshot is a read from some instant before
   * it arrived, so a push that overtook it in flight is newer — replacing
   * wholesale would resurrect a line the watcher just fired or cancelled.
   */
  async load(): Promise<void> {
    const seq = ++this.loadSeq;
    this.pushedDuringLoad.clear();
    try {
      const snapshot = await this.apiClient.chartOrders();
      if (seq !== this.loadSeq) return; // a newer load superseded this one
      const current = this.getState().orders;
      const orders = snapshot.filter((order) => !this.pushedDuringLoad.has(order.id));
      // Re-apply what the socket told us mid-flight. A pushed line that is no
      // longer in local state was retired — it must stay gone, not come back.
      for (const id of this.pushedDuringLoad) {
        const pushed = current.find((order) => order.id === id);
        if (pushed) orders.push(pushed);
      }
      this.pushedDuringLoad.clear();
      this.set({ orders, error: null });
    } catch (error) {
      if (seq !== this.loadSeq) return;
      this.set({ error: messageOf(error) });
    }
  }

  /** Clears everything — used when the trading mode changes, since practice and
   *  live lines are separate sets and must never be shown together. */
  reset(): void {
    this.firing.clear();
    this.lastPrices.clear();
    this.pushedDuringLoad.clear();
    this.loadSeq += 1; // discard any snapshot still in flight for the old account
    this.set({ orders: [], selectedId: null, error: null });
  }

  async create(draft: ChartOrderDraft): Promise<ChartOrder | null> {
    try {
      const order = await this.apiClient.createChartOrder(draft);
      this.upsert(order);
      this.set({ error: null, selectedId: order.id });
      return order;
    } catch (error) {
      this.set({ error: messageOf(error) });
      return null;
    }
  }

  /** Commits a drag. The server re-arms from the live quote, so the returned
   *  row (not the dragged-to price) is what gets stored. */
  async move(id: string, triggerPrice: number): Promise<void> {
    await this.patch(id, { triggerPrice });
  }

  async setQuantity(id: string, quantity: number): Promise<void> {
    await this.patch(id, { quantity });
  }

  /** Flips MID ↔ MKT for one line, optimistically so the pill responds instantly. */
  async toggleOrderType(id: string): Promise<void> {
    const current = this.byId(id);
    if (!current || !isWorking(current)) return;
    const orderType: ChartOrderType = current.orderType === 'mid' ? 'market' : 'mid';
    this.upsert({ ...current, orderType });
    try {
      this.upsert(await this.apiClient.updateChartOrder(id, { orderType }));
      this.set({ error: null });
    } catch (error) {
      this.upsert(current); // revert
      this.set({ error: messageOf(error) });
    }
  }

  /**
   * The ✕ on a line. A working line is cancelled server-side; a terminal one
   * (a failed fire, or one the watcher already sent) has nothing to cancel, so
   * ✕ simply dismisses it from the chart — otherwise a failed line would sit
   * there un-clearable, telling the user nothing they can act on.
   */
  async cancel(id: string): Promise<void> {
    const current = this.byId(id);
    if (current && !isWorking(current)) {
      this.remove(id);
      return;
    }
    try {
      await this.apiClient.cancelChartOrder(id);
      this.remove(id);
      this.set({ error: null });
    } catch (error) {
      // A line the watcher fired between the render and this click is no longer
      // cancellable; reload so the chart shows what actually happened.
      if (current) void this.load();
      this.set({ error: messageOf(error) });
    }
  }

  /** Server-side watcher pushed a state change over the socket. */
  applyServerUpdate(order: ChartOrder): void {
    this.firing.delete(order.id);
    this.pushedDuringLoad.add(order.id);
    if (isVisible(order)) this.upsert(order);
    else this.remove(order.id);
  }

  /**
   * Feeds a live tick in. Lines whose level the move crossed are fired through
   * the server immediately rather than waiting on the watcher's poll.
   *
   * The segment tested starts at the last price this client actually saw — or
   * the line's own `armPrice` when it has not seen one yet — so a reconnect or
   * a backgrounded tab cannot step over a level unnoticed.
   */
  applyQuote(underlying: string, last: number): void {
    if (!Number.isFinite(last) || last <= 0) return;
    const previous = this.lastPrices.get(underlying);
    this.lastPrices.set(underlying, last);

    for (const order of this.getState().orders) {
      if (order.underlying !== underlying || !isWorking(order)) continue;
      if (this.firing.has(order.id)) continue;
      const from = previous ?? order.armPrice;
      if (!chartOrderCrossed(order, from, last)) continue;
      void this.trigger(order.id);
    }
  }

  private async trigger(id: string): Promise<void> {
    this.firing.add(id);
    try {
      this.upsert(await this.apiClient.triggerChartOrder(id));
      this.set({ error: null });
    } catch (error) {
      // The watcher is still armed on this line, so a failed nudge is not the
      // end of the road — surface it and let the server path carry it.
      this.set({ error: messageOf(error) });
    } finally {
      this.firing.delete(id);
    }
  }

  private upsert(order: ChartOrder): void {
    const orders = this.getState().orders;
    const index = orders.findIndex((existing) => existing.id === order.id);
    this.set({
      orders:
        index === -1 ? [...orders, order] : orders.map((e) => (e.id === order.id ? order : e)),
    });
  }

  private remove(id: string): void {
    const { orders, selectedId } = this.getState();
    this.set({
      orders: orders.filter((order) => order.id !== id),
      selectedId: selectedId === id ? null : selectedId,
    });
  }

  private async patch(
    id: string,
    body: { triggerPrice?: number; quantity?: number },
  ): Promise<void> {
    try {
      this.upsert(await this.apiClient.updateChartOrder(id, body));
      this.set({ error: null });
    } catch (error) {
      void this.load(); // snap back to the server's view of the line
      this.set({ error: messageOf(error) });
    }
  }
}

function messageOf(error: unknown): string {
  return error instanceof Error ? error.message : String(error);
}

/** Working lines already protecting `contractSymbol` inside an OCO group. */
export function workingBracketSiblings(orders: ChartOrder[], contractSymbol: string): ChartOrder[] {
  return orders.filter(
    (order) =>
      order.contractSymbol === contractSymbol && order.ocoGroupId !== null && isWorking(order),
  );
}

/**
 * Draft for a new bracket leg (target/stop) protecting an existing position:
 * the opposite side, sized to the full position. Both legs of one position
 * share an OCO group, so filling either retires the other — the draft reuses
 * the group an existing working sibling already established, else mints a new
 * one. Shared by the chart's bracket drag and the trade-management
 * workspace's Set stop / Set target actions, so the group-reuse rule lives in
 * exactly one place.
 */
export function bracketLegDraft(input: {
  contract: OptionContract;
  position: Position;
  triggerPrice: number;
  kind: ChartOrderKind;
  orderType: ChartOrderType;
  /** All known lines; the sibling scan filters them itself. */
  orders: ChartOrder[];
}): ChartOrderDraft {
  const siblings = workingBracketSiblings(input.orders, input.position.symbol);
  return {
    underlying: input.contract.underlying,
    triggerPrice: input.triggerPrice,
    side: input.position.quantity > 0 ? 'sell' : 'buy',
    quantity: Math.abs(input.position.quantity),
    orderType: input.orderType,
    kind: input.kind,
    optionType: input.contract.optionType,
    expiration: input.contract.expiration,
    strike: input.contract.strike,
    ocoGroupId: siblings[0]?.ocoGroupId ?? crypto.randomUUID(),
  };
}

/** Short label for the kind pill: the line's colour already carries most of this. */
export function kindLabel(kind: ChartOrderKind): string {
  if (kind === 'target') return 'TP';
  if (kind === 'stop') return 'STP';
  return 'LMT';
}

/** The tappable execution pill. `MKT` is deliberately abbreviated so the pill
 *  stays the same width in both states and the row does not jump on toggle. */
export function orderTypeLabel(orderType: ChartOrderType): string {
  return orderType === 'market' ? 'MKT' : 'MID';
}

File: apps/ios/0dteTrader/Features/Chart/ChartOrders.swift
import Foundation

/// What a line is for. The colour on the chart carries this too — the label is
/// the confirmation, not the only signal.
enum ChartOrderKind: String, Codable, Sendable {
    case limit
    case target
    case stop

    var shortLabel: String {
        switch self {
        case .limit: return "LMT"
        case .target: return "TP"
        case .stop: return "STP"
        }
    }
}

enum ChartOrderStatus: String, Codable, Sendable {
    case working
    case triggered
    case filled
    case cancelled
    case failed
    case expired
}

/// A resting order line drawn on the candle chart. The broker never sees it:
/// the level is watched against the UNDERLYING, and a crossing fires a normal
/// mid/market option order.
struct ChartOrder: Equatable, Identifiable, Sendable {
    let id: String
    let underlying: String
    var triggerPrice: Double
    /// Underlying price when the line was armed. The fire test runs
    /// armPrice → last, so a gap or a relaunch cannot step over a level.
    var armPrice: Double
    let side: OrderSide
    var quantity: Int
    var orderType: ChartOrderType
    let kind: ChartOrderKind
    let optionType: OptionType
    let expiration: String
    let strike: Double
    let contractSymbol: String
    let ocoGroupId: String?
    var status: ChartOrderStatus
    let expiresAt: Date?
    var brokerOrderId: String?
    var lastError: String?

    /// Still able to fire: drawn solid, draggable, and cancellable.
    var isWorking: Bool { status == .working }

    /// Worth drawing at all. A failed line stays up so the reason is visible
    /// rather than the line silently disappearing.
    var isVisible: Bool { status == .working || status == .triggered || status == .failed }

    /// The side of the trigger this line waits to be crossed *from*.
    var armedAbove: Bool { armPrice >= triggerPrice }

    /// What the kind pill reads. A triggered line says WORKING, not SENT: an
    /// unfilled mid order is still exposed, and "sent" would read as done.
    var kindLabel: String {
        switch status {
        case .failed: return "FAILED"
        case .triggered: return "WORKING"
        default: return kind.shortLabel
        }
    }

    /// The tappable execution pill. `MKT` is abbreviated so the pill keeps its
    /// width across a toggle and the row does not jump.
    var orderTypeLabel: String { orderType.shortLabel }
}

extension ChartOrder {
    /// Nil for any unrecognised enum: a line whose side or kind we cannot read
    /// would arm a trade we cannot describe.
    init?(dto: ChartOrderDTO) {
        guard let side = OrderSide(rawValue: dto.side),
              let orderType = ChartOrderType(rawValue: dto.orderType),
              let kind = ChartOrderKind(rawValue: dto.kind),
              let optionType = OptionType(rawValue: dto.optionType),
              let status = ChartOrderStatus(rawValue: dto.status)
        else { return nil }
        self.init(
            id: dto.id,
            underlying: dto.underlying,
            triggerPrice: dto.triggerPrice,
            armPrice: dto.armPrice,
            side: side,
            quantity: dto.quantity,
            orderType: orderType,
            kind: kind,
            optionType: optionType,
            expiration: dto.expiration,
            strike: dto.strike,
            contractSymbol: dto.contractSymbol,
            ocoGroupId: dto.ocoGroupId,
            status: status,
            expiresAt: DateParsing.dateTime(dto.expiresAt),
            brokerOrderId: dto.brokerOrderId,
            lastError: dto.lastError
        )
    }
}

/// Whether a move from `previous` to `current` crosses the line's trigger from
/// the side it was armed on.
///
/// Callers pass the last price they actually observed, or the line's own
/// `armPrice` when they have not seen one — that is what makes a relaunch or a
/// backgrounded app safe. Testing the armed side (rather than a bare crossing)
/// is what stops a buy limit placed below an already-lower price from firing
/// the instant it is created.
func chartOrderCrossed(_ order: ChartOrder, previous: Double, current: Double) -> Bool {
    guard previous.isFinite, current.isFinite else { return false }
    return order.armedAbove
        ? previous >= order.triggerPrice && current <= order.triggerPrice
        : previous <= order.triggerPrice && current >= order.triggerPrice
}

/// The direction a position profits in, as a sign on the UNDERLYING.
///
/// This is why a bracket cannot simply map screen-up to "target": a long put
/// gains when the underlying falls, so its target sits BELOW the entry line and
/// its stop above. Long call → +1, long put → -1; a short position inverts.
func positionProfitDirection(optionType: OptionType, quantity: Int) -> Double {
    let isLong = quantity >= 0
    let isCall = optionType == .call
    return isLong == isCall ? 1 : -1
}

/// Whether a leg dragged to `price` from an entry at `entryPrice` is the target
/// or the stop.
func bracketKind(
    optionType: OptionType,
    quantity: Int,
    entryPrice: Double,
    price: Double
) -> ChartOrderKind {
    let profitable = (price - entryPrice) * positionProfitDirection(
        optionType: optionType,
        quantity: quantity
    ) > 0
    return profitable ? .target : .stop
}

/// Chart order lines for the signed-in account. The server is the source of
/// truth: every mutation is a request and the response replaces the local row.
///
/// Triggering is deliberately not a local order submission — when the quote
/// stream crosses a level this asks the server to fire that line, so the app
/// and the server-side watcher take the identical path and one crossing can
/// only ever produce one broker order.
@MainActor
final class ChartOrdersModel: ObservableObject {
    @Published private(set) var orders: [ChartOrder] = []
    @Published var selectedId: String?
    @Published private(set) var errorMessage: String?

    /// Chart symbol currently displayed; lines for other underlyings stay
    /// loaded but are not drawn.
    private(set) var symbol = ""

    private let apiClient: APIClient
    /// Lines with a trigger request in flight, so a burst of ticks fires once.
    private var firing: Set<String> = []
    /// Bumped per load so a superseded snapshot cannot land after a newer one.
    private var loadSeq = 0
    /// Ids the socket updated while a load was in flight. The push is strictly
    /// newer than the read that produced the snapshot, so it must win.
    private var pushedDuringLoad: Set<String> = []
    /// Last price seen per underlying — the left end of the crossing segment.
    private var lastPrices: [String: Double] = [:]

    init(apiClient: APIClient) {
        self.apiClient = apiClient
    }

    /// Lines drawn on the current chart.
    var visibleOrders: [ChartOrder] {
        orders.filter { $0.underlying == symbol && $0.isVisible }
    }

    func order(id: String) -> ChartOrder? {
        orders.first { $0.id == id }
    }

    func setSymbol(_ symbol: String) {
        guard symbol != self.symbol else { return }
        self.symbol = symbol
        selectedId = nil
    }

    /// Re-reads the account's lines. Called on appear, on every socket
    /// reconnect, and on foreground — pushes that landed while the stream was
    /// down are gone for good.
    ///
    /// Merges rather than replaces. A snapshot is a read from some instant
    /// before it arrived, so a push that overtook it in flight is newer;
    /// replacing wholesale would resurrect a line the watcher just fired.
    func load() async {
        loadSeq += 1
        let seq = loadSeq
        pushedDuringLoad.removeAll()
        do {
            let snapshot = try await apiClient.chartOrders().compactMap(ChartOrder.init(dto:))
            guard seq == loadSeq else { return } // a newer load superseded this one
            let current = orders
            var merged = snapshot.filter { !pushedDuringLoad.contains($0.id) }
            // Re-apply what the socket told us mid-flight. A pushed line that is
            // no longer in local state was retired — it stays gone.
            for id in pushedDuringLoad {
                if let pushed = current.first(where: { $0.id == id }) { merged.append(pushed) }
            }
            pushedDuringLoad.removeAll()
            orders = merged
            errorMessage = nil
        } catch let error as APIError {
            guard seq == loadSeq else { return }
            errorMessage = error.userMessage
        } catch {
            guard seq == loadSeq else { return }
            errorMessage = error.localizedDescription
        }
    }

    /// Clears everything — used when the trading mode changes, since practice
    /// and live lines are separate sets and must never be shown together.
    func reset() {
        firing.removeAll()
        lastPrices.removeAll()
        pushedDuringLoad.removeAll()
        loadSeq += 1 // discard any snapshot still in flight for the old account
        orders = []
        selectedId = nil
        errorMessage = nil
    }

    @discardableResult
    func create(_ draft: ChartOrderDraftDTO) async -> ChartOrder? {
        do {
            guard let order = ChartOrder(dto: try await apiClient.createChartOrder(draft)) else {
                errorMessage = "The server returned an order this build cannot read."
                return nil
            }
            upsert(order)
            selectedId = order.id
            errorMessage = nil
            Haptics.success()
            return order
        } catch let error as APIError {
            errorMessage = error.userMessage
            Haptics.error()
            return nil
        } catch {
            errorMessage = error.localizedDescription
            return nil
        }
    }

    /// Commits a drag. The server re-arms from the live quote, so the returned
    /// row — not the dragged-to price — is what gets stored.
    func move(id: String, triggerPrice: Double) async {
        await patch(id: id, ChartOrderPatchDTO(triggerPrice: triggerPrice))
    }

    /// Flips MID ↔ MKT for one line, optimistically so the pill responds at once.
    func toggleOrderType(id: String) async {
        guard let current = order(id: id), current.isWorking else { return }
        let next: ChartOrderType = current.orderType == .mid ? .market : .mid
        var optimistic = current
        optimistic.orderType = next
        upsert(optimistic)
        Haptics.selection()
        do {
            guard let updated = ChartOrder(
                dto: try await apiClient.updateChartOrder(
                    id: id,
                    patch: ChartOrderPatchDTO(orderType: next.rawValue)
                )
            ) else { return }
            upsert(updated)
            errorMessage = nil
        } catch let error as APIError {
            upsert(current) // revert
            errorMessage = error.userMessage
        } catch {
            upsert(current)
            errorMessage = error.localizedDescription
        }
    }

    /// The ✕ on a line. A working line is cancelled server-side; a terminal one
    /// (a failed fire, or one the watcher already sent) has nothing to cancel,
    /// so ✕ simply dismisses it — otherwise a failed line would sit there
    /// un-clearable, telling the user nothing they can act on.
    func cancel(id: String) async {
        if let current = order(id: id), !current.isWorking {
            remove(id: id)
            Haptics.impact(.light)
            return
        }
        do {
            try await apiClient.cancelChartOrder(id: id)
            remove(id: id)
            errorMessage = nil
            Haptics.impact(.light)
        } catch let error as APIError {
            // A line the watcher fired a moment ago is no longer cancellable;
            // reload so the chart shows what happened, not a stale line.
            errorMessage = error.userMessage
            await load()
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    /// Server-side watcher pushed a state change over the socket.
    func applyServerUpdate(_ order: ChartOrder) {
        firing.remove(order.id)
        pushedDuringLoad.insert(order.id)
        if order.isVisible {
            upsert(order)
        } else {
            remove(id: order.id)
        }
    }

    /// Feeds a live tick in, returning the lines whose level it crossed. The
    /// caller fires them; the segment tested starts at the last price this app
    /// actually saw — or the line's own arm price when it has seen none — so a
    /// relaunch or a backgrounded app cannot step over a level unnoticed.
    func crossedOrders(underlying: String, last: Double) -> [ChartOrder] {
        guard last.isFinite, last > 0 else { return [] }
        let previous = lastPrices[underlying]
        lastPrices[underlying] = last
        return orders.filter { order in
            order.underlying == underlying
                && order.isWorking
                && !firing.contains(order.id)
                && chartOrderCrossed(order, previous: previous ?? order.armPrice, current: last)
        }
    }

    /// Asks the server to fire a line now rather than waiting on the watcher's
    /// next poll. Idempotent server-side, so racing the watcher is harmless.
    func trigger(id: String) async {
        guard !firing.contains(id) else { return }
        firing.insert(id)
        defer { firing.remove(id) }
        do {
            if let updated = ChartOrder(dto: try await apiClient.triggerChartOrder(id: id)) {
                upsert(updated)
            }
            errorMessage = nil
        } catch let error as APIError {
            // The watcher is still armed on this line, so a failed nudge is not
            // the end of the road — surface it and let the server path carry it.
            errorMessage = error.userMessage
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func clearError() {
        errorMessage = nil
    }

    // MARK: - Private

    private func patch(id: String, _ body: ChartOrderPatchDTO) async {
        do {
            if let updated = ChartOrder(dto: try await apiClient.updateChartOrder(id: id, patch: body)) {
                upsert(updated)
            }
            errorMessage = nil
        } catch let error as APIError {
            errorMessage = error.userMessage
            await load() // snap back to the server's view of the line
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    private func upsert(_ order: ChartOrder) {
        if let index = orders.firstIndex(where: { $0.id == order.id }) {
            orders[index] = order
        } else {
            orders.append(order)
        }
    }

    private func remove(id: String) {
        orders.removeAll { $0.id == id }
        if selectedId == id { selectedId = nil }
    }
}
